Adding persistence to task settings dialog system

skyler-rs opened the pull request:
Task tree "Edit Settings" entries now render as buttons and will dynamically load a settings dialog for the parent Task.

Current settings are still test values and aren't integrated with the respective tasks.

If no project is loaded, editing task settings will persist for the session only(currently).

If a project has been loaded, settings will be saved in the project folder under [projectName].settings/ as json files.

When a project is loaded or created, current settings are cleared, system defaults are reloaded and project settings are loaded (see below).

When a project is loaded, all .json files found in the project folder [projectName].settings/ anr loaded and merged into the current settings json.

KochynVolodymyr commented on src/Main/Settings.cpp:
```diff
@@ -24,10 +24,24 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
 using namespace FOEDAG;
 
+#define SETTINGS_DEBUG false
```
you can use qDebug() << ""; Looks weird.

skyler-rs replied:
These are rare use debug prints that are only needed when actively developing in this area so i don't want the prints coming out everytime someone is running debug.

KochynVolodymyr commented on src/Main/WidgetFactory.cpp:
```diff
@@ -26,11 +26,33 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
 const QString DlgBtnBoxName{"SettingsDialogButtonBox"};
 
+// Local debug helpers
+#define WIDGET_FACTORY_DEBUG false
```
why not qDebug() ?

skyler-rs replied:
These are rare use debug prints that are only needed when actively developing in this area so i don't want the prints coming out everytime someone is running debug.

KochynVolodymyr commented on src/Compiler/TaskTableView.cpp:
```diff
@@ -47,11 +47,27 @@ TaskTableView::TaskTableView(TaskManager *tManager, QWidget *parent)
+    if (parent != nullptr) {
+      // Handle the event first so the selection moves before the modal dialog
+      QTableView::mousePressEvent(event);
+      handled = true;
```
instead of this flag you can just return after emit.

skyler-rs replied:
Sounds good, I try to avoid mid function returns out of habit, but if that's standard in our project I'll switch over to that.

KochynVolodymyr replied:
Not sure about standard but we have many returns in the middle of the function

KochynVolodymyr commented on src/Compiler/TaskTableView.cpp:
```diff
@@ -47,11 +47,27 @@ TaskTableView::TaskTableView(TaskManager *tManager, QWidget *parent)
+  auto task = m_taskManager->task(idx.row());
+  bool handled = false;
+
+  if (task && idx.data(Qt::DisplayRole).toString() == "Edit settings") {
```
Can you please add one more field to Task - something like Type
emun TaskType {Action, Settings, ...}. 
In this **if** check the task type but not the text.

skyler-rs replied:
Will do

KochynVolodymyr commented on src/Compiler/TaskTableView.cpp:
```diff
@@ -104,7 +120,16 @@ QRect TaskTableView::expandArea(const QModelIndex &index) const {
                               const QStyleOptionViewItem &option,
                               const QModelIndex &index) const {
-  if (index.data(ParentDataRole).toBool()) {
+  if (index.data(Qt::DisplayRole).toString() == "Edit settings") {
```
same here with text

skyler-rs commented on src/Compiler/TaskTableView.cpp:
```diff
@@ -47,11 +47,27 @@ TaskTableView::TaskTableView(TaskManager *tManager, QWidget *parent)
+      // Handle the event first so the selection moves before the modal dialog
+      QTableView::mousePressEvent(event);
+      handled = true;
+      emit TaskDialogRequested(parent->title());
```
This was just trying to dynamically work with the existing tableview code. I can look into your suggestions below about a settings type and look into providing a specific category when creating the setting.

KochynVolodymyr replied:
I'm just curious what happens if we change title? As I understand then we have to change something in the settings and maybe something else. I don't think we have to be depend on the title anywhere. this will cause the problem in the future.

skyler-rs replied:
If you changed the parent category title and didn't update the settings.json then it would load an empty json object and display an empty dialog. We can either get it all free dynamically and expect the dev to update their meta data or explicitly define lookup values and require the dev to define those at the beginning. Since i'll be creating a TaskType now I can probably associate that Task w/ a fixed string so that it would survive the dev renaming the categories down the road.

KochynVolodymyr replied:
okay

skyler-rs commented:
Addressed review comments. Please note the following:
For ChildItemDelegate::paint(), there is no access to the task manager unless I add it to the constructor so instead I'm adding  a new TaskTypeRole. I'm currently converting from enum to uint before storing or after reading to avoid adding Q_NAMESPACE
Q_ENUM_NS(TaskType) to the FOEDAG namespace.

Regarding calling setIndexWidget instead of drawing a button in the ChildDelegate paint, I didn't see a good place to call that since the table is mostly constructed from the view. It seems like i'd need to manually scan the tree to find the right indexes to add the button to. Let me know if I'm missing something though.

alain-rs commented:
@skyler-rs, please resolve conflict with @KochynVolodymyr PR

skyler-rs commented:
@alain-rs Conflict resolved and the final test run should be finishing shortly
